lib/paypal_express/api.rb

```diff
@@ -187,16 +199,43 @@ def reset_payment_methods(kb_account_id, payment_methods, properties, context)
 
       def build_form_descriptor(kb_account_id, descriptor_fields, properties, context)
         # Pass extra parameters for the gateway here
-        options           = {}
-        properties        = merge_properties(properties, options)
-
-        # Add your custom static hidden tags here
-        options           = {
-            #:token => config[:paypal-express][:token]
-        }
-        descriptor_fields = merge_properties(descriptor_fields, options)
+        options              = {}
+        properties           = merge_properties(properties, options)
+        descriptor_fields    = merge_properties(descriptor_fields, options)
+        form_fields          = properties_to_hash(descriptor_fields)
+        descriptor           = super(kb_account_id, descriptor_fields, properties, context)
+        kb_account           = @kb_apis.account_user_api.get_account_by_id(kb_account_id, context)
+        kb_payment_method_id = (@kb_apis.payment_api.get_account_payment_methods(kb_account_id, false, [], context).find { |pm| pm.plugin_name == 'killbill-paypal-express' }).id
+        amount               = (form_fields[:amount] || "0").to_i
+        currency             = form_fields[:currency] || kb_account.currency
+
+        response = @private_api.initiate_express_checkout(kb_account_id,
+                                                          context.tenant_id,
+                                                          amount,
+                                                          currency,
+                                                          false,
+                                                          form_fields)
+        unless response.success?
+          raise "Unable to initiate paypal express checkout: #{response.message}"
+        end
 
-        super(kb_account_id, descriptor_fields, properties, context)
+        descriptor.form_url   = @private_api.to_express_checkout_url(response, context.tenant_id, options)
+
+        properties_hash       = properties_to_hash(properties)
+        should_create_payment = Killbill::Plugin::ActiveMerchant::Utils.normalized(properties_hash, :create_pending_payment)
+
+        if should_create_payment
+          @kb_apis.payment_api.create_purchase(kb_account,
+                                               kb_payment_method_id,
+                                               nil,
+                                               amount,
+                                               currency,
+                                               nil,
+                                               response.token,
```

Comment:
Reasonable defaults but we should make the `paymentExternalKey` and `paymentTransactionExternalKey` configurable nonetheless, so the client can set them.